### PR TITLE
[framework] add methods to create additional mint/burn capabilities

### PR DESCRIPTION
**Pull request by mkurnikov**
## Motivation

In some cases, it could be useful to have many Mint/Burn capabilities for one coin, as there could be different contracts that need to mint or burn. For example, one contract could be a Staking contract (which needs the mint capability for inflation and burn capability for slashing) and another contract could be a Vesting contract, which mints new coins for investors, early adopters, and team members.
With current PR we bring an approach allowing us to create new capabilities using existing ones, it's just a concept for now, but could be nice to discuss.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

I've added a test that invokes the new functionality.

### CODE REVIEW

File: aptos-move/framework/aptos-framework/sources/Coin.move [hunk 1]
```diff
@@ -263,6 +263,16 @@ module AptosFramework::Coin {
         Coin<CoinType> { value: amount }
     }
 
+    /// Allows to create multiple `MintCapability` resources for the coins.
+    public fun add_extra_mint_capability<CoinType>(_cap: &MintCapability<CoinType>): MintCapability<CoinType> {
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/Coin.move:267]: nit: May be name this create_mint_capability()?

File: aptos-move/framework/aptos-framework/sources/Coin.move [hunk 2]
```diff
@@ -263,6 +263,16 @@ module AptosFramework::Coin {
         Coin<CoinType> { value: amount }
     }
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/Coin.move:265]: If we're adding ways to create mint/burn caps, we should probably add ways to destroy them as well (they're not droppable).

**Review by movekevin — COMMENTED** [review]
There's a potential design consideration here: Should creating a coin with multiple minters/burns be done with the Coin module, or should be it be a wrapper module such as ManagedCoin? The trade off here is complexity vs functionality - if we keep adding more functionalities, Coin can quickly become very complex.

**Comment by borispovod** [comment]
> There's a potential design consideration here: Should creating a coin with multiple minters/burns be done with the Coin module, or should be it be a wrapper module such as ManagedCoin? The trade off here is complexity vs functionality - if we keep adding more functionalities, Coin can quickly become very complex.

There are several options on how we can design it IMHO. Theoretically, we can get a rid from many mint/burn abilities and use only friends functions in new deployed modules contains coin generic type, while the devs would store mint/burn cap indeed under their accounts, yet in my opinion it's still breaking language design and makes it hard to correctly work with Move (because caps should be stored somewhere, usually it's coin dev address, and there should be option to extract caps when some friend module calling mint/burn functions, it seems brakes Move logic a little, because you will have to hardcore address where caps stored in the code, not sure it's good design). 

So i think option with many capabilities could an option (in the same time i still think we should do something with `Coin::initialize` and get rid of `Reflect` there).

**Comment by davidiw** [comment]
@borispovod 

> So i think option with many capabilities could an option (in the same time i still think we should do something with `Coin::initialize` and get rid of `Reflect` there).

Could you elaborate? Is this about TypeInfo?

**Review by davidiw — CHANGES_REQUESTED** [review]
These capabilities are already copy. I think there's a better solution in what we already have in the Capability.move from the standard library, but I think there are a lot of nuances here... Should a capability be copy, should it allow further delegation, and how do you protect it from misuse.

I'm not sure what this PR is adding.

**Comment by borispovod** [comment]
> @borispovod
> 
> > So i think option with many capabilities could an option (in the same time i still think we should do something with `Coin::initialize` and get rid of `Reflect` there).
> 
> Could you elaborate? Is this about TypeInfo?

Yeah, let's talk in our chat. I will explain what i mean.

Regarding the current PR, it's my mistake, i totally forgot about copy ability as it's not very popular approach :) 

@mkurnikov can you close the current PR?

**Pull request closed without merging**